### PR TITLE
Add a generic SubjectPublicKeyInfo decoder

### LINKED ISSUE #15646: Decoding SubjectPublicKeyInfo structures is very inefficient
I have the following file (a DH public key):

````
-----BEGIN PUBLIC KEY-----
MIIBIDCBlQYJKoZIhvcNAQMBMIGHAoGBAKO3UUHeLGtYuOeAefepPlI9wgw+mzQx
/UA2TSyGDj1iu6U7MsOwYzcRa0+dsI/UMeS1Q/gf+g6scoyk6MelTcfPxEb9xZ4S
6+cqsy+8ExZMxTgnM5LfjGEDcNB4qLGKgM4Ryag+azEhg6rvR+xBHOa5cVGpTay2
2VYsqnUl7EjjAgECA4GFAAKBgQCeFtXoVpN1mQ8cSvBoZzFDtcuzzSp6wdo5vU8I
OU6xgr1UgvdEJxtgeBeL5OKu/Z6wavtSOX565GmFvGob78F1kpyJMiCk5mvDm2XJ
5p29enY2n0HAPBHzQNgYDdEoqlH3MWK3pnAkGbFxepp3Vaqg6/KL3IrQ6FULr/71
0IVtIQ==
-----END PUBLIC KEY-----
````

Using the program below I can read it:

````
#include <stdio.h>
#include <openssl/pem.h>
#include <openssl/evp.h>

int main(void)
{
    BIO *bio = BIO_new_file("dhpubkey.pem", "r");
    EVP_PKEY *pkey = PEM_read_bio_PUBKEY(bio, NULL, NULL, NULL);

    if(pkey == NULL)
        printf("Failed\n");
    else
        printf("Success\n");

    EVP_PKEY_free(pkey);

    return 0;
}
````

The above code will eventually (via the decoders in a provider) call this function - which actually reads a SubjectPublicKeyInfo encoded DH key:

https://github.com/openssl/openssl/blob/6d2e0076e6a8146f3bdaa60151b2fa6e4df0d1bf/crypto/dh/dh_ameth.c#L60-L109

By setting a breakpoint in this function, I see that this decoding routine is actually called 11 times for my simple program above! The key is successfully decoded 10 times and then thrown away. On the 11th time it is successfully decoded and kept.

What happens is this:

1) The PEM file is first decoded to SubjectPublicKeyInfo  (SPKI) DER data
2) There are 11 built-in decoders that know handle SPKI data (one for each key type).
3) The first one gets called (which is for RSA). This uses d2i_RSA_PUBKEY to attempt to read the RSA key. Internally this calls d2i_X509_PUBKEY. That function detects that the key is DH and calls the correct DH decoding function to successfully read the key. The read key is then passed back up the stack to d2i_RSA_PUBKEY which checks its type, discovers that it isn't an RSA key and so throws it away and fails
4) Since the first SPKI decoder failed we move on to the second one which is for RSA-PSS. This is implemented in the same way, and so the DH key is decoded successfully again and then thrown away.
5) We continue to iterate through all the different key types which are all implemented in the same way. So we repeatedly successfully decode the DH key and then throw it away.
6) The final key type in the list to check is a DH key. This time it actually works and the key is returned.

### PR DESCRIPTION
**Pull request by mattcaswell**
Previously all the SubjectPublicKeyInfo decoders were specific to a key
type. We would iterate over all them until a match was found for the correct
key type. Each one would fully decode they key before then testing whether
it was a match or not - throwing it away if not. This was very inefficient.

Instead we introduce a generic SubjectPublicKeyInfo decoder which figures
out what type of key is contained within it, before subsequently passing on
the data to a key type specific SubjectPublicKeyInfo decoder.

Fixes #15646

### CODE REVIEW

**Comment by mattcaswell** [comment]
This is marked as draft because it includes all the commits from #15498.

File: crypto/x509/x_pubkey.c [hunk 1]
```diff
@@ -64,6 +64,22 @@ ASN1_SEQUENCE(X509_PUBKEY_INTERNAL) = {
...
 } static_ASN1_SEQUENCE_END_name(X509_PUBKEY, X509_PUBKEY_INTERNAL)
 
+X509_PUBKEY *ossl_d2i_X509_PUBKEY_INTERNAL(const unsigned char **pp,
+                                           long len)
+{
+    X509_PUBKEY *xpub = OPENSSL_zalloc(sizeof(*xpub));
+
+    if (xpub == NULL)
+        return NULL;
+    return (X509_PUBKEY *)ASN1_item_d2i((ASN1_VALUE **)&xpub, pp, len,
```
Comment (slontis) [inline crypto/x509/x_pubkey.c:74]: ASN1_item_d2i_ex() ?

File: providers/implementations/encode_decode/endecoder_common.c [hunk 2]
```diff
@@ -82,3 +84,18 @@ void ossl_prov_free_key(const OSSL_DISPATCH *fns, void *key)
         kmgmt_free(key);
 }
 
+int ossl_read_der(PROV_CTX *provctx, OSSL_CORE_BIO *cin,  unsigned char **data,
+                  long *len)
+{
+    BUF_MEM *mem = NULL;
+    BIO *in = ossl_bio_new_from_core_bio(provctx, cin);
+    int ok = (asn1_d2i_read_bio(in, &mem) >= 0);
```
Comment (slontis) [inline providers/implementations/encode_decode/endecoder_common.c:92]: This potentially raises an error here...
Should it use ERR_set_mark() or is this done at a higher level?

File: providers/implementations/encode_decode/decode_spki2typespki.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    if (!ossl_read_der(ctx->provctx, cin, &der, &len))
+        return 1;
+    derp = der;
+    xpub = ossl_d2i_X509_PUBKEY_INTERNAL((const unsigned char **)&derp, len);
+
+    /* We return "empty handed".  This is not an error. */
+    if (xpub == NULL
+            || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algor, xpub)) {
+        X509_PUBKEY_free(xpub);;
```
Comment (slontis) [inline providers/implementations/encode_decode/decode_spki2typespki.c:72]: ;;

File: providers/implementations/encode_decode/decode_spki2typespki.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        return 1;
+    }
+    X509_ALGOR_get0(&oid, NULL, NULL, algor);
+
+    if (!OBJ_obj2txt(dataname, sizeof(dataname), oid, 0))
+        return 0;
```
Comment (slontis) [inline providers/implementations/encode_decode/decode_spki2typespki.c:78]: Wont this return leak?

File: providers/implementations/encode_decode/decode_spki2typespki.c [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    ossl_X509_PUBKEY_INTERNAL_free(xpub);
```
Comment (slontis) [inline providers/implementations/encode_decode/decode_spki2typespki.c:80]: Why do you call this here to free - but up above you use X509_PUBKEY_free()?

File: providers/implementations/encode_decode/endecoder_common.c (hunk 2 above)
Comment (mattcaswell) [reply providers/implementations/encode_decode/endecoder_common.c:92]: Its done higher up

File: providers/implementations/encode_decode/decode_spki2typespki.c (hunk 5 above)
Comment (mattcaswell) [reply providers/implementations/encode_decode/decode_spki2typespki.c:88]: Should be `ossl_X509_PUBKEY_INTERNAL_free` in both places.

**Comment by mattcaswell** [comment]
Feedback above addressed and rebased. Also had to add a fixup for dealing with some SM2 test failures (not sure why I wasn't getting those test failures originally!)

Please take another look.

File: providers/implementations/encode_decode/decode_spki2typespki.c [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        return 1;
+    derp = der;
+    xpub = ossl_d2i_X509_PUBKEY_INTERNAL((const unsigned char **)&derp, len,
+                                         PROV_LIBCTX_OF(ctx->provctx));
+
+    /* We return "empty handed".  This is not an error. */
+    if (xpub == NULL
+            || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algor, xpub)) {
+        ossl_X509_PUBKEY_INTERNAL_free(xpub);
+        return 1;
```
Comment (levitte) [inline providers/implementations/encode_decode/decode_spki2typespki.c:75]: `xpub == NULL` is the "empty handed" return, I agree with that.  Not so sure a failing `X509_PUBKEY_get0_param()`, I think I'd categorise that as an error that would stop the decoding process.  So strictly speaking:

``` C
    if (xpub == NULL)
        return 1;
    if (!!X509_PUBKEY_get0_param(NULL, NULL, NULL, &algor, xpub)) {
        ossl_X509_PUBKEY_INTERNAL_free(xpub);
        return 0;
    }
```

**Review by levitte — CHANGES_REQUESTED** [review]

File: providers/implementations/encode_decode/decode_spki2typespki.c (hunk 6 above)
Comment (mattcaswell) [reply providers/implementations/encode_decode/decode_spki2typespki.c]: Fixed

**Comment by mattcaswell** [comment]
Updated commit addressing feedback pushed. Also addressed a number of CI failures.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Good work!

File: providers/implementations/encode_decode/decode_spki2typespki.c [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+#ifndef OPENSSL_NO_EC
+    /* SM2 abuses the EC oid, so this could actually be SM2 */
+    if (OBJ_obj2nid(oid) == NID_X9_62_id_ecPublicKey
+            && ossl_x509_algor_is_sm2(algor)) {
+        strcpy(dataname, "SM2");
+    } else
+#endif
+    if (!OBJ_obj2txt(dataname, sizeof(dataname), oid, 0)) {
+        goto end;
+    }
```
Comment (levitte) [inline providers/implementations/encode_decode/decode_spki2typespki.c:92]: Why do you have braces around either one-statement body?

**Comment by mattcaswell** [comment]
Fixup pushed to address the code style nit raised by @levitte. I am assuming that doesn't affect the approval.

**Comment by t8m** [comment]
> I am assuming that doesn't affect the approval.

Sure, it's still good to be merged.

**Comment by mattcaswell** [comment]
Pushed. Thanks!

**Pull request closed without merging**